### ISSUE TITLE
Issue with SF Candidate Contacts

### ISSUE THREAD

**Issue by sadatmalik** [talentbeyondboundaries/tbbtalentv2#981]
From Sarah:

I was trying to update the no. dependants in these 2 candidates opportunities but I am getting an error when I try to save because the candidate contact isn't linked in from the TC, whats the best way to link in the candidate contact without messing anything up? :sweat_smile:

1. <https://talentbeyondboundaries.lightning.force.com/lightning/r/Opportunity/006Uu0000051g3xIAA/view>
2. <https://talentbeyondboundaries.lightning.force.com/lightning/r/Opportunity/006Uu000005Fx6YIAS/view>

[Slack Message](https://refugeejobsmarket.slack.com/archives/C7QK72T4J/p1715603743638959?thread_ts=1715603743.638959&cid=C7QK72T4J)

**Comment by cazcam34** [comment]
Reason these candidate contact records can't be linked to the candidate opp is that their candidate contact records are missing TBB Ids. Digging into the SF backend I found that this candidate contact search dropdown is populated by candidate contacts that have TBB ids greater than 0 (see screenshot).
<img width="1041" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBN1hmQkE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--7b00230d0fc4b645c4e7f0ddef7de2d6139e04f4/Screen%20Shot%202024-05-15%20at%203.31.25%20pm.png" alt="Screen Shot 2024-05-15 at 3.31.25 pm.png" />

See Slack response with explanation:
Hi Yousef looking this further, the reason the candidate doesn't appear in the Candidate Contact search is that the candidate record (here) doesn't have a TBB Id. That is what is used to determine which candidates appear in the search (candidate contact TBB Id field is greater than 0).
Candidate SF contacts are automatically created when the candidate is added to a submission list on the TC (unless it already exists from previous opportunities), and if done this way the TBB ID should be automatically populated as well. Can I ask how you created this candidate's contact record? Was it via the TC or did you create it on SF directly?
Sarah Walder this is also connected to you not being able to update the candidate opportunities. Those candidate opps are associated with candidate contacts (here & here) that are missing TBB Ids, therefore they cannot be linked to the Candidate Opportunities as they don't appear in the Candidate Contact search.
In terms of updating these two records to fix your issues, I don't have access on Salesforce to edit the TBB Id fields so may need higher level access. 
Sadat Malik are you able to edit the candidate contact's TBB Ids on Salesforce?

**Comment by cazcam34** [comment]
Waiting on Sam's return so we can edit these candidate contacts with TBB Ids. Waiting to hear back from Yousef on how he created his contact, I assume it was done directly on SF hence why it isn't properly populated but can't say for sure yet. If he did create directly on SF, the other contact created by Rana was probably done the same way as it's similar. If needed we could do a post or a video on how the TC is integrated with SF and how contacts, opps, jobs are created via the TC.

**Comment by samschlicht** [comment]
Hey @cazcam34 hadn't seen this issue and response yet — appears to be the same with the second contact that Sarah flags, this time created by Rana. I've restated your question to Yousef and also to Rana, so we'll see what comes back. I don't think we should resolve the issue till we better understand the cause.

Strangely, Yousef's SF contact was linked to in the TC. I don't know how he could have done that without creating it from there in the first place! Again, we probably need to rule out a TC bug here.

Noting separately that we should probably all be SF admins and this is something on my and Sadat's 1-1 list.

**Comment by samschlicht** [comment]
At this point, we have no evidence to treat this as a TC issue — it appears that TBB users manually created their own SF candidate contacts and as such the records are incomplete and causing issues. For that reason, I'm continuing this work on the Asana Ops Tech board. I will rule out all possibilities and make a new issue on here if there turns out to be anything TC-related going on.

**Comment by samschlicht** [comment]
For posterity, the issue was this:

- The SF candidate contacts had not been successfully created by the TC.
- I haven't been able to get to the bottom of why, except to say that our candidate creation code is now passing the usual manual tests (including for these two candidates), which leads me to assume either that something went wrong during those individual SF API calls.
- Perfectly understandably, Yousef and Rana attempted to manually create these contacts, but were not able to enter the TBB ID (which is locked to all but admins) and therefore couldn't find them from the lookup search in the candidate opportunity records.

Our steps to fix this are not to enter the TBB ID on the manually created contacts, but instead to do the following:

- Delete the manually created SF contact(s), which will be easy to identify, since they'll have been created by another TBB user than John, who is the TC's proxy user. They will also not have the usual candidate contact layout, since that's locked to all but SF admins. 
- On the TC, navigate to the candidate profile > Additional Info tab and first delete the SF link, if any (Yousef had actually added a link to his manually created SF contact), then hit the 'Create/update SF' button.
- Then add the newly created candidate contact to the relevant candidate opp via the lookup search field.

I've done a quick search and found these to be the only manually created SF candidate contacts, so I'm actually fairly confident we don't need to take any further steps to avoid this user behaviour, or to log these steps, but do let me know if you think otherwise!

**Comment by samschlicht** [comment]
This was my user reply in the SF channel:
@Yousef @Sarah Walder

Both these SF opportunities now have the correct candidate linked. Apologies for the delay but I wanted to look into it properly before resolving.

1. For all our SF users: if you ever find that a SF candidate contact record you expected to see (typically because you’ve added the candidate to a submission list) appears to be missing, there are two steps to take:
2. On the TC, navigate to the candidate’s profile > ‘Additional Info’ tab > ‘Create/update Salesforce’ button. If that step doesn’t work, contact the TC/SF team (it will usually be me who handles these requests, but others can jump in if I’m unavailable)

Manually entering candidates in SF will not work, and is liable to cause further issues — though it’s a perfectly understandable thing to try!
